File: packages/sporewright/src/trust.ts
// SPDX-License-Identifier: LGPL-3.0-only WITH LGPL-3.0-linking-exception
/**
 * trust — the pure cross-device trust / freshness / churn math.
 *
 * This is a PURE, side-effect-free, storage-agnostic module: no SQL, no
 * `Date.now()`, no `Tensor` mutation. It is deliberately NOT a `Tensor` method:
 * the pairwise-increment model is a mutation keyed on a device PAIR, not a level
 * roll-up / reduce, so it does not belong on the Rust-twinned, golden-vector-locked
 * `Tensor` class. It lives here as free functions.
 *
 * The math is relocated VERBATIM from `jobcache/shared/src/fact-tree.ts` (the live
 * pairwise model, jobcache 2026-06). Numbers are byte-identical to that source — the
 * adapter (fact-tree.ts) keeps every DB read/write and delegates the arithmetic here.
 * The DB stays the authority on the persisted value (same contract as `Tensor`: the
 * package exposes math only; the DB versions the write).
 */

// --- Trust constants (relocated verbatim from fact-tree.ts) ---
export const TRUST_AGREE_BASE_GAIN = 0.02;
export const TRUST_DISAGREE_BASE_LOSS = 0.04;
export const TRUST_TIME_HALF_LIFE_HOURS = 48;
/** numberValue(..., 0.5) fallback used when a device has no stored trust_score yet. */
export const TRUST_DEFAULT = 0.5;
/** State-machine thresholds (the SQL CASE in updateDeviceTrust). */
export const TRUST_TRUSTED_AT = 0.8;
export const TRUST_LIMITED_BELOW = 0.3;

// --- Freshness / re-investigation / churn constants (relocated verbatim) ---
export const VERIFY_CONFLICT_MIN_WEIGHT = 0.1;
export const FACT_CHURN_COOLDOWN_THRESHOLD = 6;
export const FACT_CHURN_WINDOW_MS = 86_400_000;
export const FACT_COOLDOWN_MS = 86_400_000;

/**
 * Clamp to [0, 1]. Verbatim from fact-tree.ts:560-563.
 *
 * NOTE: a non-finite input snaps to 0.5 (the trust default), NOT to 0 or 1. This is
 * load-bearing — a "cleanup" to plain `Math.max(0, Math.min(1, x))` would silently
 * change trust behaviour when a bad (e.g. Infinity) value leaks in. Copy as-is.
 */
export function clamp01(value: number): number {
  if (!Number.isFinite(value)) return 0.5;
  return Math.max(0, Math.min(1, value));
}

/**
 * Time-decayed evidence weight: closer observations = stronger evidence.
 *
 * Pure form of fact-tree.ts:565-571. The adapter computes the time delta (it owns the
 * `Date.parse` / `Date` handling) and passes `deltaHours` already in hours. A non-finite
 * `deltaHours` (e.g. an unparseable timestamp upstream) returns 0.5, mirroring the
 * source's `if (!Number.isFinite(...)) return 0.5` fallback — keep that fallback in BOTH
 * the adapter's delta calc AND here.
 */
export function timeWeight(deltaHours: number): number {
  if (!Number.isFinite(deltaHours)) return 0.5;
  return Math.pow(2, -Math.abs(deltaHours) / TRUST_TIME_HALF_LIFE_HOURS);
}

export interface PairTrustResult {
  incomingTrust: number;
  currentTrust: number;
}

/**
 * Symmetric pairwise trust increment. Each device's trust moves by the base
 * gain/loss weighted by the OTHER device's current trust and by the freshness weight.
 *
 * Behaviour-identical to fact-tree.ts:2416-2426: `base = agree ? +GAIN : -LOSS`, then
 * `next = clamp01(self + base * other * freshness)` for each side.
 */
export function updatePairTrust(input: {
  incomingTrust: number;
  currentTrust: number;
  agree: boolean;
  freshness: number;
}): PairTrustResult {
  const base = input.agree ? TRUST_AGREE_BASE_GAIN : -TRUST_DISAGREE_BASE_LOSS;
  return {
    incomingTrust: clamp01(input.incomingTrust + base * input.currentTrust * input.freshness),
    currentTrust: clamp01(input.currentTrust + base * input.incomingTrust * input.freshness),
  };
}

export type TrustState =
  | "trusted"
  | "limited"
  | "suspended"
  | "revoked"
  | "probationary"
  | string;

/**
 * Trust state-machine reduction — pure form of the SQL CASE in
 * updateDeviceTrust fact-tree.ts:2446-2451.
 *
 * 'suspended' and 'revoked' are sticky terminal states; otherwise cross the
 * thresholds: >= TRUST_TRUSTED_AT → 'trusted', < TRUST_LIMITED_BELOW → 'limited',
 * else unchanged. (Not yet bound in the relocate — the adapter keeps the SQL CASE
 * authoritative; this is the pure twin for tests / future convergence.)
 */
export function nextTrustState(prev: TrustState, nextScore: number): TrustState {
  if (prev === "suspended" || prev === "revoked") return prev;
  if (nextScore >= TRUST_TRUSTED_AT) return "trusted";
  if (nextScore < TRUST_LIMITED_BELOW) return "limited";
  return prev;
}

/**
 * Freshness-driven re-investigation gate — pure form of shouldQueueFollowUp
 * fact-tree.ts:2320-2329. The adapter supplies the already-resolved churn count, the
 * cooldown deadline as epoch-ms (or null), and `nowMs` (it owns `Date.now()`).
 */
export function shouldReinvestigate(input: {
  freshness: number;
  churnCount: number;
  cooldownUntilMs: number | null;
  nowMs: number;
}): boolean {
  if (input.freshness < VERIFY_CONFLICT_MIN_WEIGHT) return false;
  if (input.churnCount >= FACT_CHURN_COOLDOWN_THRESHOLD) return false;
  return input.cooldownUntilMs == null || input.cooldownUntilMs <= input.nowMs;
}

/**
 * Task priority for a re-investigation — pure form of the literal
 * `Math.max(1, Math.round(1000 * freshness))` at fact-tree.ts:2376.
 */
export function reinvestigationPriority(freshness: number): number {
  return Math.max(1, Math.round(1_000 * freshness));
}

/**
 * Provenance of a churn timestamp output. `'carried'` means the value is the prior
 * stored timestamp unchanged — the adapter should re-emit its stored value VERBATIM
 * (the live model carried these via `dateString(...)`, preserving the exact stored
 * string, not a re-canonicalized one). `'fresh'` means the value is newly computed from
 * the observation time and the adapter serializes `valueMs`. `'observed'` means the value
 * is the observation time itself (used as the window fallback when there is no prior
 * window-start). This discriminant exists purely so the adapter reproduces byte-identical
 * bound parameters; the arithmetic decision (which source each output comes from) is made
 * here in the pure function.
 */
export interface ChurnTimestamp {
  /** Numeric epoch-ms value of this timestamp (null only for an absent cooldown). */
  valueMs: number | null;
  /** Where the value came from — drives byte-identical serialization in the adapter. */
  source: "carried" | "fresh" | "observed";
}

export interface ChurnState {
  count: number;
  /** Epoch-ms of the window start (always present). */
  windowStartedAtMs: number;
  /** Epoch-ms of the cooldown deadline, or null when none is set. */
  cooldownUntilMs: number | null;
  /** Provenance of windowStartedAtMs — `'carried'` | `'fresh'` | `'observed'`. */
  windowStartedAt: ChurnTimestamp;
  /** Provenance of cooldownUntilMs — `'carried'` (incl. null carry) | `'fresh'`. */
  cooldownUntil: ChurnTimestamp;
}

/**
 * Churn-window roll/reset/cooldown — pure form of nextChurn fact-tree.ts:2294-2318.
 *
 * The adapter owns all timestamp parsing and serialization (Date ↔ epoch-ms). It passes
 * the previous state as epoch-ms (or null if the field has no prior state) and the new
 * observation time as epoch-ms; this returns the next state in epoch-ms PLUS a provenance
 * discriminant per timestamp so the adapter can serialize byte-identically to the live
 * model (carried values re-emit the stored string verbatim; fresh values serialize the ms;
 * the window fallback uses the observation time).
 *
 * - `same` (the new observation matches the stored value): hold the existing window;
 *   carry prior count / window-start / cooldown unchanged (count defaults to 0, the
 *   window-start falls back to the observation time, cooldown to null when there is no
 *   prior state — matching `dateString(...) ?? observedAt.toISOString()` / `?? null`).
 * - changed: if the prior window is still active (within FACT_CHURN_WINDOW_MS), increment
 *   the count and keep the window start (re-emitted fresh, as the live model did with
 *   `windowStart.toISOString()`); otherwise start a fresh window of count 1. When the count
 *   reaches FACT_CHURN_COOLDOWN_THRESHOLD, set a cooldown FACT_COOLDOWN_MS out (fresh);
 *   otherwise carry the prior cooldown forward verbatim.
 */
export function nextChurn(input: {
  prev: {
    count: number;
    /**
     * Parsed epoch-ms of the stored window-start (null if the stored value is absent OR
     * unparseable). Drives the active-window arithmetic — mirrors the live model's
     * `dateValue(...)` in the changed branch.
     */
    windowStartedAtMs: number | null;
    /**
     * Whether a stored window-start string/Date is PRESENT (truthy), regardless of
     * parseability. Drives verbatim carry in the `same` branch — mirrors the live model's
     * `dateString(...)` (a non-empty unparseable string is carried verbatim, not parsed).
     */
    hasStoredWindow: boolean;
    cooldownUntilMs: number | null;
  } | null;
  same: boolean;
  observedAtMs: number;
}): ChurnState {
  const { prev, same, observedAtMs } = input;

  if (same) {
    // windowStartedAt: a present stored window-start is carried verbatim (even if
    // unparseable), else the observation time — matching `dateString(...) ?? observedAt`.
    const window: ChurnTimestamp = prev?.hasStoredWindow
      ? { valueMs: prev.windowStartedAtMs, source: "carried" }
      : { valueMs: observedAtMs, source: "observed" };
    return {
      count: prev?.count ?? 0,
      windowStartedAtMs: window.valueMs ?? observedAtMs,
      cooldownUntilMs: prev?.cooldownUntilMs ?? null,
      windowStartedAt: window,
      cooldownUntil: { valueMs: prev?.cooldownUntilMs ?? null, source: "carried" },
    };
  }

  const windowStartMs = prev?.windowStartedAtMs ?? null;
  const activeWindow = windowStartMs != null && observedAtMs - windowStartMs <= FACT_CHURN_WINDOW_MS;
  const count = activeWindow ? (prev?.count ?? 0) + 1 : 1;
  // The live model re-canonicalizes the active window start (`windowStart.toISOString()`)
  // and uses observedAt for a fresh window, so both are 'fresh' (serialize the ms).
  const startedAtMs = activeWindow ? windowStartMs! : observedAtMs;
  const cooldownMs = count >= FACT_CHURN_COOLDOWN_THRESHOLD ? observedAtMs + FACT_COOLDOWN_MS : prev?.cooldownUntilMs ?? null;
  return {
    count,
    windowStartedAtMs: startedAtMs,
    cooldownUntilMs: cooldownMs,
    windowStartedAt: { valueMs: startedAtMs, source: "fresh" },
    cooldownUntil:
      count >= FACT_CHURN_COOLDOWN_THRESHOLD
        ? { valueMs: cooldownMs, source: "fresh" }
        : { valueMs: cooldownMs, source: "carried" },
  };
}

File: crates/sporewright/src/trust.rs
// SPDX-License-Identifier: LGPL-3.0-only WITH LGPL-3.0-linking-exception
//! Pure trust, freshness, and churn arithmetic.
//!
//! The host owns clocks, persistence, identities, and evidence collection. This
//! module only transforms explicit inputs, which keeps it useful in centralized
//! orchestrators and peer-local routers alike.

pub const TRUST_AGREE_BASE_GAIN: f64 = 0.02;
pub const TRUST_DISAGREE_BASE_LOSS: f64 = 0.04;
pub const TRUST_TIME_HALF_LIFE_HOURS: f64 = 48.0;
pub const TRUST_DEFAULT: f64 = 0.5;
pub const TRUST_TRUSTED_AT: f64 = 0.8;
pub const TRUST_LIMITED_BELOW: f64 = 0.3;
pub const VERIFY_CONFLICT_MIN_WEIGHT: f64 = 0.1;
pub const FACT_CHURN_COOLDOWN_THRESHOLD: u64 = 6;
pub const FACT_CHURN_WINDOW_MS: i64 = 86_400_000;
pub const FACT_COOLDOWN_MS: i64 = 86_400_000;

/// Clamp to `[0, 1]`. Non-finite evidence resets to the neutral default.
pub fn clamp01(value: f64) -> f64 {
    if !value.is_finite() {
        return TRUST_DEFAULT;
    }
    value.clamp(0.0, 1.0)
}

/// Exponential freshness weight with a 48-hour half-life.
pub fn time_weight(delta_hours: f64) -> f64 {
    if !delta_hours.is_finite() {
        return TRUST_DEFAULT;
    }
    2_f64.powf(-delta_hours.abs() / TRUST_TIME_HALF_LIFE_HOURS)
}

#[derive(Clone, Copy, Debug, PartialEq)]
pub struct PairTrustResult {
    pub incoming_trust: f64,
    pub current_trust: f64,
}

/// Symmetric pairwise trust update.
///
/// `source_volatility` is clamped to `[0, 1]` and discounts disagreement only:
/// a frequently changing source is weak evidence that either observer is wrong.
/// Agreement is unchanged. Passing `0` preserves the original model exactly.
pub fn update_pair_trust(
    incoming_trust: f64,
    current_trust: f64,
    agree: bool,
    freshness: f64,
    source_volatility: f64,
) -> PairTrustResult {
    let volatility = clamp01(source_volatility);
    let base = if agree {
        TRUST_AGREE_BASE_GAIN
    } else {
        -TRUST_DISAGREE_BASE_LOSS * (1.0 - volatility)
    };
    PairTrustResult {
        incoming_trust: clamp01(incoming_trust + base * current_trust * freshness),
        current_trust: clamp01(current_trust + base * incoming_trust * freshness),
    }
}

/// Sticky terminal-state reduction plus the public score thresholds.
pub fn next_trust_state(previous: &str, next_score: f64) -> &str {
    if matches!(previous, "suspended" | "revoked") {
        previous
    } else if next_score >= TRUST_TRUSTED_AT {
        "trusted"
    } else if next_score < TRUST_LIMITED_BELOW {
        "limited"
    } else {
        previous
    }
}

pub fn should_reinvestigate(
    freshness: f64,
    churn_count: u64,
    cooldown_until_ms: Option<i64>,
    now_ms: i64,
) -> bool {
    freshness >= VERIFY_CONFLICT_MIN_WEIGHT
        && churn_count < FACT_CHURN_COOLDOWN_THRESHOLD
        && (cooldown_until_ms.is_none()
            || cooldown_until_ms.is_some_and(|deadline| deadline <= now_ms))
}

pub fn reinvestigation_priority(freshness: f64) -> i64 {
    (1_000.0 * freshness).round().max(1.0) as i64
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum TimestampSource {
    Carried,
    Fresh,
    Observed,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct ChurnTimestamp {
    pub value_ms: Option<i64>,
    pub source: TimestampSource,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct PreviousChurn {
    pub count: u64,
    pub window_started_at_ms: Option<i64>,
    pub has_stored_window: bool,
    pub cooldown_until_ms: Option<i64>,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct ChurnState {
    pub count: u64,
    pub window_started_at_ms: i64,
    pub cooldown_until_ms: Option<i64>,
    pub window_started_at: ChurnTimestamp,
    pub cooldown_until: ChurnTimestamp,
}

/// Roll, reset, or cool down a host-supplied churn window.
pub fn next_churn(previous: Option<PreviousChurn>, same: bool, observed_at_ms: i64) -> ChurnState {
    if same {
        let window = match previous {
            Some(previous) if previous.has_stored_window => ChurnTimestamp {
                value_ms: previous.window_started_at_ms,
                source: TimestampSource::Carried,
            },
            _ => ChurnTimestamp {
                value_ms: Some(observed_at_ms),
                source: TimestampSource::Observed,
            },
        };
        let cooldown = previous.and_then(|value| value.cooldown_until_ms);
        return ChurnState {
            count: previous.map_or(0, |value| value.count),
            window_started_at_ms: window.value_ms.unwrap_or(observed_at_ms),
            cooldown_until_ms: cooldown,
            window_started_at: window,
            cooldown_until: ChurnTimestamp {
                value_ms: cooldown,
                source: TimestampSource::Carried,
            },
        };
    }

    let active_window = previous
        .and_then(|value| value.window_started_at_ms)
        .is_some_and(|started| observed_at_ms - started <= FACT_CHURN_WINDOW_MS);
    let count = if active_window {
        previous.map_or(1, |value| value.count + 1)
    } else {
        1
    };
    let started_at_ms = if active_window {
        previous
            .and_then(|value| value.window_started_at_ms)
            .expect("active window has a start")
    } else {
        observed_at_ms
    };
    let fresh_cooldown = count >= FACT_CHURN_COOLDOWN_THRESHOLD;
    let cooldown = if fresh_cooldown {
        Some(observed_at_ms + FACT_COOLDOWN_MS)
    } else {
        previous.and_then(|value| value.cooldown_until_ms)
    };
    ChurnState {
        count,
        window_started_at_ms: started_at_ms,
        cooldown_until_ms: cooldown,
        window_started_at: ChurnTimestamp {
            value_ms: Some(started_at_ms),
            source: TimestampSource::Fresh,
        },
        cooldown_until: ChurnTimestamp {
            value_ms: cooldown,
            source: if fresh_cooldown {
                TimestampSource::Fresh
            } else {
                TimestampSource::Carried
            },
        },
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn volatility_discounts_only_disagreement() {
        let stable = update_pair_trust(0.5, 0.5, false, 1.0, 0.0);
        let volatile = update_pair_trust(0.5, 0.5, false, 1.0, 0.75);
        let fully_volatile = update_pair_trust(0.5, 0.5, false, 1.0, 1.0);
        assert_eq!(stable.incoming_trust, 0.48);
        assert_eq!(volatile.incoming_trust, 0.495);
        assert_eq!(fully_volatile.incoming_trust, 0.5);
        assert_eq!(
            update_pair_trust(0.5, 0.5, true, 1.0, 1.0),
            update_pair_trust(0.5, 0.5, true, 1.0, 0.0)
        );
    }

    #[test]
    fn freshness_and_churn_match_the_published_constants() {
        assert_eq!(time_weight(48.0), 0.5);
        let previous = PreviousChurn {
            count: 5,
            window_started_at_ms: Some(1_000),
            has_stored_window: true,
            cooldown_until_ms: None,
        };
        let next = next_churn(Some(previous), false, 2_000);
        assert_eq!(next.count, 6);
        assert_eq!(next.cooldown_until_ms, Some(2_000 + FACT_COOLDOWN_MS));
    }
}
